File: 2023/ts/day_4_part_1.ts
import fs from "fs";

const file = String(fs.readFileSync("./day_4_input.txt"));

let sum = 0;
for (const line of file.split("\n").slice(0, -1)) {
  const numbers = line.split(": ")[1];

  const winningNumbers = numbers
    .split(" | ")[0]
    .split(" ")
    .filter((s) => s !== "")
    .map((n) => parseInt(n));
  const yourNumbers = numbers
    .split(" | ")[1]
    .split(" ")
    .filter((s) => s !== "")
    .map((n) => parseInt(n));

  let points = 0;
  for (const yourNumber of yourNumbers) {
    if (winningNumbers.includes(yourNumber)) {
      if (points === 0) points++;
      else points *= 2;
    }
  }

  sum += points;
}

console.log(sum);

File: 2023/python/day_4_part_1.py
def main():
    with open("./day_4_input.txt", "r") as f:
        lines = f.readlines()

        sum = 0
        for line in lines:
            line = line.strip()

            numbers = line.split(": ")[1]

            winning_numbers = [int(n) for n in numbers.split(" | ")[0].split(" ") if n]
            your_numbers = [int(n) for n in numbers.split(" | ")[1].split(" ") if n]

            count = -1
            for num in your_numbers:
                if num in winning_numbers:
                    count += 1

            points = 2**count if count >= 0 else 0
            sum += points

    print(sum)


if __name__ == "__main__":
    main()
